cassiopeia/solver/distance_correction/_crispr_cas9_distance_correction_solver.py

```diff
@@ -0,0 +1,556 @@
+    )
+    height = _inverse(
+        f=expected_hamming_distance_given_height,
+        y=observed_hamming_distance,
+        lower=0,
+        upper=1,
+    )
+    estimated_tree_distance = 2 * height
+    return estimated_tree_distance
+
+
+def _ternary_hamming_distance(
+    s1: List[int],
+    s2: List[int],
+    missing_state_indicator: int = -1,
+) -> float:
+    """
+    (Scaled) ternary Hamming distance, ignoring missing data.
+
+    Here, by `scaled` we mean that the Hamming distance is divided by the total
+    number of characters considered, thus bringing it into the range [0, 2].
+
+    Here, `ternary` means that we score two mutated states that are different
+    as 2 rather then 1.
+
+    Args:
+        s1: Character states of first leaf
+        s2: Character states of second leaf.
+        missing_state_indicator: Missing data indicator.
+
+    Returns:
+        (Scaled) ternary Hamming distance between s1 and s2, ignoring missing
+        data.
+    """
+    dist = 0
+    num_present = 0
+    for i in range(len(s1)):
+        if s1[i] == missing_state_indicator or s2[i] == missing_state_indicator:
+            continue
+        num_present += 1
+        if s1[i] != s2[i]:
+            if s1[i] > 0 and s2[i] > 0:
+                dist += 2
+            else:
+                dist += 1
+        else:
+            dist += 0
+    if num_present == 0:
+        return 0
+    return dist / num_present
+
+
+def _crispr_cas9_expected_ternary_hamming_distance(
+    height: float,
+    mutation_proportion: float,
+    collision_probability: float,
+) -> float:
+    """
+    Expected ternary Hamming distance under the CRISPR-Cas9 model.
+
+    Expected ternary Hamming distance function for two leaves with LCA at a
+    height of `height` in an ultrametric tree of depth 1.
+
+    Args:
+        height: The height of the LCA of the two leaves.
+        mutation_proportion: Mutation proportion parameter of the model.
+        collision_probability: Collision probability parameter of the model.
+
+    Returns:
+        The expected ternary Hamming distance.
+    """
+    h = height
+    exp_minus_r = 1.0 - mutation_proportion
+    q = collision_probability
+    if exp_minus_r < 1e-8:
+        # To avoid doing log(0) later on.
+        return 0.0
+    else:
+        r = -np.log(exp_minus_r)  # The estimate of the mutation rate.
+        return (
+            2 * (1 - np.exp(-h * r)) ** 2 * (1 - q)
+            + 2 * (1 - np.exp(-h * r)) * (np.exp(-h * r))
+        ) * (np.exp(r * (h - 1)))
+
+
+def crispr_cas9_corrected_ternary_hamming_distance(
+    mutation_proportion: float,
+    collision_probability: float,
+    s1: List[int],
+    s2: List[int],
+    missing_state_indicator: int,
+) -> float:
+    """
+    Corrected ternary Hamming distance.
+
+    The "corrected" distance is an estimate of true tree distance.
+
+    It is assumed that the tree is ultrametric of depth 1.
```

Comment:
If we keep "ternary" as the name, I would suggest defining it here too.

Reply:
Added.